### DISCUSSION TITLE
Optimising repeat cache reads in graphcache

### DISCUSSION THREAD

**Discussion by nathanmarks** [urql-graphql/urql#2722] [Q&A]
Hey there,

We've been looking for alternatives to apollo because their normalised cache performance has been causing us problems. Urql graphcache looks significantly better, especially in terms of write performance.

There is an observation I've made though, and I'm wondering if this is something that can be optimised or if it's difficult with the cache design: 

Assuming a cache containing data to fulfill a query and no operations updating the cache data, I can repeatedly mount a `useQuery` with apollo and only pay the cache read cost once. So if I mount it 50 times, only the 1st mount incurs the full cost of a cache read, the repeated mounts are all super cheap. With urql, there's no such optimisation, and the full cache read price is paid repeatedly even if the cache is untouched (nothing dirty at all, not just for data relevant to the query in question).

Is there any way to optimise this? Is it possible to do in the graphcache design? or is it a design compromise made in other to achieve other objectives

thanks!

**Comment by JoviDeCroock** [comment]
Hey and welcome to urql!

We have been working on the referrential equality of objects and arrays but generally we have been focussing a lot on the read-write performance. We might have regressed in the equality but we do have a bunch of tests confirming that it should work 😅 maybe the root-object is reset in certain integrations. Do you have examples where this happens?

**Reply by nathanmarks** [reply]
Thanks!

Let me try put together a codesandbox today to show you what I mean. In the meantime, can you point me to the specs that cover this?

Thanks!

**Reply by JoviDeCroock** [reply]
What do you mean with the specs, you want the code? 😅 this is the key-value map that we use to keep refs https://github.com/FormidableLabs/urql/blob/main/exchanges/graphcache/src/cacheExchange.ts#L54

**Reply by nathanmarks** [reply]
Is this where things are going wrong? https://github.com/FormidableLabs/urql/blob/b9ca5d3e39565bdc5dff61bd66e6e629ec3618ce/exchanges/graphcache/src/operations/query.ts#L100-L107

The repeated cost I'm trying to avoid seems to be in `readSelection`, or should it not have gotten this far to begin with?

**Reply by JoviDeCroock** [reply]
yeah, should not get that far, we have a fix for that in the works https://github.com/FormidableLabs/urql/pull/2718 do note that generally the cache itself is a lot faster than most others (in benchmarks at least) so wasteful reads should not really be a problem... Are you seeing any main thread blocking or?

**Reply by nathanmarks** [reply]
not seeing blocking, no. And yeah, the actual read is faster than the read from other caches. But I'd like to be able to put a best foot forward in the benchmarks that I'm sharing with the team (which include repeated cache reads between unmounting/mounting components, not just perf of a single cache read).

At the moment, using a mocked fetch that resolves with the response immediately, my benchmarks doing 100 runs with a cold cache are faster than those with a warm cache. Which definitely isn't expected/ideal.

If I apply [these changes](https://github.com/FormidableLabs/urql/commit/da121d45512477844a651cf94da4854009adcb9d) locally, should I see the issue fixed?

**Reply by JoviDeCroock** [reply]
You should atleast see an improvement, the improvement should also be seen when adding `__typename` to the selection-sets though

**Reply by nathanmarks** [reply]
Hmmm... I made those changes, still seeing it reach `readSelection`.

Is there a console.log I can chuck in somewhere to see where something is being missed?

([here's](https://www.diffchecker.com/iSFXbjgI) the changes I made btw)

**Reply by nathanmarks** [reply]
I also just realised, the query I'm testing with already has `__typename` on all the fields.

**Reply by nathanmarks** [reply]
> Is this where things are going wrong?
> 
> https://github.com/FormidableLabs/urql/blob/b9ca5d3e39565bdc5dff61bd66e6e629ec3618ce/exchanges/graphcache/src/operations/query.ts#L100-L107
> 
> The repeated cost I'm trying to avoid seems to be in `readSelection`, or should it not have gotten this far to begin with?

@JoviDeCroock you mentioned it shouldn't get this far -- where should it get to? I can work from there and debug what's happening in my scenario
